Refactoring ocean temperature - plus other new ocean diagnostics and preprocessors.

ledm opened the pull request:
Duplicating the pull request #436  https://github.com/ESMValGroup/ESMValTool/pull/436#issuecomment-398441199
due to the pul request coming from a personal fork. 

After several excellent contributions from various members of the community, an initial draft of the ocean diagnostic section is now available for review. 

This pull request contains the following new features:
- A fix for latitudes values greater than 90 degrees in HadGEM2-ES and HadGEM2-CC, described in issue #378 
- A new folder called `ocean` in the `esmvaltool/diag_scripts`. This folder contains four new diagnostics and one toolkit. The four new diagnostics are:
  - diagnostic_timeseries.py: Produces a simple time series plot
  - diagnostic_maps.py: Produces a 2D map of the region
  - diagnostic_transects.py: Produces an image of the transects
  - diagnostic_profiles.py: Produces a profile plot
- The toolkit is diagnostic_tools.py, and it contains several useful functions used by the four ocean diagnostics.

Two new name list files are also included in this pull request:
-  esmvaltool/namelists/namelist_OceanBGC.yml:
-  esmvaltool/namelists/namelist_OceanPhysics.yml
These two name lists are not meant to be fully functioning nameliests to evaluate all facets of ocean models.  Instead, they should serve as an example of how to produce various kinds of figure of the ocean. They also make use of several preprocessors. Note that these diagnostics are not a direct replacement for the previous marine metrics (namelist_GlobalOcean.yml), described in issue #350. 

In addition, several new preprocessors were needed in order to evaluate of CMIP5 models marine components. 
- average_volume
- time_average
- extract_volume 
- extract_trajectory
- extract_slice
- depth_integration

Finally, two new vertical interpolation schemes were added:
- nearest_extrap 
- linear_extrap

bouweandela commented:
Please merge the `version2_development` branch into this branch to fix the tests.

bouweandela commented on esmvaltool/preprocessor/_multimodel.py:
```diff
@@ -73,7 +73,10 @@ def _compute_statistic(datas, name):
         # datas is per time point
         # so we can safely NOT compute stats for single points
-        u_datas = [data for data in datas if not np.all(data.mask)]
+        if datas.ndim == 1:
```
@valeriupredoi Can you comment on this? Does it work as intended?

ledm replied:
Just to add some context, this section was added because I ran the `time_average` preprocessor followed by the multi model mean. The time average preprocessor produces a 1D array, which was not among the range of possibilities considered by the multimodel preprocessor.

mattiarighi replied:
Time operations are still not part of the preprocessor, although a draft module `_time_area.py` is available. These kind of operations should anyway be performed at the end of the preprocessing chain, i.e. after level selection, regrdding, fill value masking and multi-model-stats. This is the way the preprocessing chain is currently designed, since a customizable order of operations is not yet available (altough planned). 

Hence my suggestion here would be to move the `time_average` at the end of the chain.

valeriupredoi replied:
this should work OK, I've just tested with this branch as well and basic preprocessing works a-ok

ledm replied:
@mattiarighi, I've added several functions to the preprocessor, and I think I've added all of them into the _time_area.py module. However, very few of them are time manipulations. Would it be better to split this into several files with different categories, ie, time, area, volume?

bouweandela commented:
Please use the PEP8 conventions for naming functions and variables:
https://www.python.org/dev/peps/pep-0008/#prescriptive-naming-conventions
Specifically (see the Codacy issues, click the Details link above behind Codacy)

> Function names should be lowercase, with words separated by underscores as necessary to improve readability.
>
> Variable names follow the same convention as function name

Please use the numpy docstring style, at the very least for the functions that are part of the preprocessor:
http://sphinxcontrib-napoleon.readthedocs.io/en/latest/example_numpy.html

koldunovn commented on esmvaltool/preprocessor/_time_area.py:
```diff
@@ -112,15 +159,185 @@ def area_average(mycube, coord1, coord2):
+    # CMOR ised data should already have bounds?
+    #    mycube.coord(coord1).guess_bounds()
+    #    mycube.coord(coord2).guess_bounds()
     grid_areas = iris.analysis.cartography.area_weights(mycube)
```
Do I understand it right that you guessing area weights? For curvilinear grids, it's not the best strategy since the error might be quite substantial. I think it is much clearer to use `fx` files, provided by modelling groups,  in particular, `areacello`. The information about the location of this files is going to be available  when this issue is solved #423

ledm replied:
I believe that iris does attempt to guess the area weights from the grid boundaries in the coordinates. 

If there is a method in ESMValTool to do that from the provided area files, then that would be an improvement. 

Better still would be for iris to load the fx files alongside the cube, so that the methods in cartograhpy like area_weights work as intended.

bouweandela replied:
Maybe you could open an issue on the [iris github page](https://github.com/SciTools/iris/issues) to request such a feature.

koldunovn replied:
At the end of the day, I would compare results that you get by using `area_weights` and by applying the `areacello` directly for some particularly crazy grids. If you are happy with the result, probably it is safe to leave it as is for now.

ledm replied:
Back to @bouweandela's original comment:
>Specifically (see the Codacy issues, click the Details link below behind Codacy)

I feel pretty silly sending in regular work in progress reports trying to figure out what codacy wants from me. It is possible to run that locally, or is it hard to set up? the yapf/isort utilities don't seem to agree with the code standards used in codacy. 

Also, with regards to the codacy report, do I need to address all the issues, or just the yellow or higher issues? Ie, I can't make codacy able to import numpy (blue warning)!

jvegasbsc replied:
You can run prospector (https://prospector.landscape.io/en/master/) locally as Codacy issues are mostly based on prospector warnings. This greatly speed up the process 

I think you can install it trough conda easily

valeriupredoi replied:
either that or ```sudo apt-get install``` if on Debian; from a low level point of view I'd add a check if the data already has bounds, ```iris``` throws an error if it already has

bouweandela replied:
@ledm Indeed it is recommended that you run prospector locally, as described in our README.md file: https://github.com/esmvalgroup/esmvaltool/tree/version2_development#code-style

bouweandela commented:
It would also be nice if you could add unit tests for the new preprocessor functions.

jvegasbsc commented on esmvaltool/diag_scripts/ocean/diagnostic_tools.py:
```diff
@@ -0,0 +1,176 @@
+
+
+def sensibleUnits(cube, name):
+    """
```
Instead of forcing the units, it will be nice to have them as a configuration in the diagnostic

ledm commented on esmvaltool/cmor/_fixes/CMIP5/HadGEM2_ES.py:
```diff
@@ -0,0 +1,52 @@
+        # print("Fixing HadGEM2-CC Latitude metadata")
+        lat = cube.coord('latitude')
+        lat.points = np.clip(lat.points, -90., 90.)
+        lat.bounds = np.clip(lat.bounds, -90., 90.)
```
I don't think that they mean anything. 
The problem looks like a floating precision issue, as many of the latitude coordinates are only slightly away from their nearby interger: ie 90.00000763

koldunovn replied:
They are meaningless, it is safe to remove them.

jvegasbsc replied:
ok, perfect

jvegasbsc commented on esmvaltool/diag_scripts/ocean/diagnostic_tools.py:
```diff
@@ -0,0 +1,176 @@
+        name = '/'.join(name)
+    if name[-1] != '/':
+        name = name + '/'
+    if os.path.exists(name) is False:
```
Use * to unpack the list

```python
a = os.path.join(*['a','b','c'])
```

I was not talking about the part that is creating the path, just the composition of it

bouweandela commented:
@jvegasbsc @ledm
>I have one general comment about the pull request: your diagnostics can be useful for more than just the ocean. After all, plotting a map, a section or a timeseries is done for variables in all the modelling realms. Also, I don see any assumption in the implementation that could make them fail for atmos variables, for example
>
>It does not really make sense to hide them inside an 'ocean' folder. I think they belong to a 'general' or 'basic' folder so more people can benefit from your work

Generally useful stuff should go into `esmvaltool/diag_scripts/shared`, if it is plotting related in the subdirectory `plot`.

ledm commented on esmvaltool/namelists/namelist_OceanPhysics.yml:
```diff
@@ -0,0 +1,500 @@
+      Regional_Ocean_mean_profile: &Regional_Ocean_mean_profile
+        script: ocean/diagnostic_profiles.py
+        
+                                            
```
As far as I'm aware, the regridding was not implemented for curvilinear grids, as per (issue #230)[https://github.com/ESMValGroup/ESMValTool/issues/230]. In fact, the subject of how best to interpolate grid is still very much under debate between @valeriupredoi, @tillku, myself and others. 

I have tested with the irregular grids, but they fail (no surprise there). In addition, some of the new preprocessors are not yet adapted to irregular grids, like extract_trajectory and extract_slice, and I will need to revisit those preprocessors when irregular grid regridding is available.

ledm commented:
@jvegasbsc @bouweandela 

>>I have one general comment about the pull request: your diagnostics can be useful for more than just the ocean. After all, plotting a map, a section or a timeseries is done for variables in all the modelling realms. Also, I don see any assumption in the implementation that could make them fail for atmos variables, for example
>>It does not really make sense to hide them inside an 'ocean' folder. I think they belong to a 'general' or 'basic' folder so more people can benefit from your work

>Generally useful stuff should go into esmvaltool/diag_scripts/shared, if it is plotting related in the subdirectory plot.

I'm in favour of putting these tools as shared/general tools, however I feel that the current range of ocean diagnostics are still fairly basic in their capability. In the future, it is entirely possible that there will be much more ocean specific parts needed, and it would be good to keep that option available.

valeriupredoi commented:
sorry - I meant only the reversal of changes from ```_time_area.py```

ledm commented:
I've split the _time_area.py preprocessor into three separate files, and updated preprocessor/_init__.py accordingly. 

The parts that I wrote and added seem to have passed their codacy tests, but a recent merge with version2_development has caused the CircleCI tests to fail the  test_no_namelist and the test_avoid_casing_collisions tests.

bouweandela commented on esmvaltool/preprocessor/__init__.py:
```diff
@@ -14,16 +14,16 @@
-from ._time_area import extract_slice as extract_slice
-from ._time_area import depth_integration as depth_integration
-from ._time_area import seasonal_mean
+from ._time_pp import time_average as time_average
```
There's no need for the `as time_average`, as you are using the same name, so this can be removed.

bouweandela commented:
Yes, that test is there to warn you that you still have the namelists folder with namelists in it that you need to rename to recipe:
https://github.com/ESMValGroup/ESMValTool/tree/REFACTORING_OceanTemperature/esmvaltool/namelists

mattiarighi commented:
Ok. I've just asked because I've seen there are some changes in `_multimodel.py` to account for different dimensionality in input and I was thinking this was due to some area/time operations performed in advance.

valeriupredoi commented:
hey @ledm great work in getting all the codacy/lint-changes done and cheers for devving this whole thing! I will take a careful look at the PR myself today but I don't think I can come up with anything new that @mattiarighi @bouweandela or @jvegasbsc haven't suggested yet. In the meantime, could you pls point me to a receipe that I can run after I have installed the branch just to test the functionality? Cheers! ...and as a general remark, let's try with smaller and more incremental PR's next time, this one's a monster, but a good monster :grin:

ledm commented:
@mattiarighi I think that the multi-model operations always occur towards the end of the preproccessor, after any time or area operations. (rightly so, too)

The changes were needed in `_multimodel.py` because a zero-dimensional scenario (time dimensional 1D arrays) had not yet been considered. 

Thanks @valeriupredoi, The [recipe_OceanPhysics.yml](https://github.com/ESMValGroup/ESMValTool/blob/REFACTORING_OceanTemperature/esmvaltool/recipes/recipe_OceanPhysics.yml) recipe should be a good starting point. Let me know if there's anything else I can do.

mattiarighi commented:
Well, not really: the order of operations was chosen to minimize the loss of information through the different steps. Performing multi-model stats after the area/time operations will return something different, and in general less accurate, than applying that before. The same goes for regridding and masking of fill values.

In the end this is of course a choice of the diagnostic developer, but since we are supporting only one order of operation (so far), I believe we should choose it in a way that gives the best accuracy (even if this is not the most efficient in terms of cpu/memory).

This approach has also the advantage that, being more general, allows to apply the same preprocessor set to multiple diagnostics.

valeriupredoi commented:
@ledm cheers, I grabbed the recipe and installed the branch: two mentions -- one not so important:

line ```# Takahashi, T. et al., Mar. Chem., 164, 95–125,``` in the namelist contains a special character at position 1040 that will make yaml crash, have a look

and another, rather important: your plotting will NOT work with iris 1.13 because of switches within ```nc_time_axis``` to cf_units (this a nested problem starting from iris.quickplot and I have the same issue with my autoassess plotting; the solution is either to not use quickplot or to return an empty plot if iris 1.13 is used). See stack:
```
Traceback (most recent call last):
  File "/home/users/valeriu/anaconda3/envs/esmvaltool_v2_dev/lib/python3.6/site-packages/ESMValTool-2.0a0-py3.6.egg/esmvaltool/diag_scripts/ocean/diagnostic_timeseries.py", line 267, in <module>
    main(config)
  File "/home/users/valeriu/anaconda3/envs/esmvaltool_v2_dev/lib/python3.6/site-packages/ESMValTool-2.0a0-py3.6.egg/esmvaltool/diag_scripts/ocean/diagnostic_timeseries.py", line 248, in main
    metadatas,
  File "/home/users/valeriu/anaconda3/envs/esmvaltool_v2_dev/lib/python3.6/site-packages/ESMValTool-2.0a0-py3.6.egg/esmvaltool/diag_scripts/ocean/diagnostic_timeseries.py", line 188, in multi_model_time_series
    lw=2.,
  File "/home/users/valeriu/anaconda3/envs/esmvaltool_v2_dev/lib/python3.6/site-packages/iris/quickplot.py", line 268, in plot
    result = iplt.plot(*args, **kwargs)
  File "/home/users/valeriu/anaconda3/envs/esmvaltool_v2_dev/lib/python3.6/site-packages/iris/plot.py", line 1113, in plot
    return _draw_1d_from_points('plot', _plot_args, *args, **kwargs)
  File "/home/users/valeriu/anaconda3/envs/esmvaltool_v2_dev/lib/python3.6/site-packages/iris/plot.py", line 607, in _draw_1d_from_points
    result = draw_method(u, v, *args, **kwargs)
  File "/home/users/valeriu/anaconda3/envs/esmvaltool_v2_dev/lib/python3.6/site-packages/matplotlib/pyplot.py", line 3358, in plot
    ret = ax.plot(*args, **kwargs)
  File "/home/users/valeriu/anaconda3/envs/esmvaltool_v2_dev/lib/python3.6/site-packages/matplotlib/__init__.py", line 1855, in inner
    return func(ax, *args, **kwargs)
  File "/home/users/valeriu/anaconda3/envs/esmvaltool_v2_dev/lib/python3.6/site-packages/matplotlib/axes/_axes.py", line 1528, in plot
    self.add_line(line)
  File "/home/users/valeriu/anaconda3/envs/esmvaltool_v2_dev/lib/python3.6/site-packages/matplotlib/axes/_base.py", line 1932, in add_line
    self._update_line_limits(line)
  File "/home/users/valeriu/anaconda3/envs/esmvaltool_v2_dev/lib/python3.6/site-packages/matplotlib/axes/_base.py", line 1954, in _update_line_limits
    path = line.get_path()
  File "/home/users/valeriu/anaconda3/envs/esmvaltool_v2_dev/lib/python3.6/site-packages/matplotlib/lines.py", line 951, in get_path
    self.recache()
  File "/home/users/valeriu/anaconda3/envs/esmvaltool_v2_dev/lib/python3.6/site-packages/matplotlib/lines.py", line 651, in recache
    xconv = self.convert_xunits(self._xorig)
  File "/home/users/valeriu/anaconda3/envs/esmvaltool_v2_dev/lib/python3.6/site-packages/matplotlib/artist.py", line 191, in convert_xunits
    return ax.xaxis.convert_units(x)
  File "/home/users/valeriu/anaconda3/envs/esmvaltool_v2_dev/lib/python3.6/site-packages/matplotlib/axis.py", line 1526, in convert_units
    ret = self.converter.convert(x, self.units, self)
  File "/home/users/valeriu/anaconda3/envs/esmvaltool_v2_dev/lib/python3.6/site-packages/nc_time_axis/__init__.py", line 277, in convert
    raise ValueError('The datetime attribute of the CalendarDateTime '
ValueError: The datetime attribute of the CalendarDateTime object must be of type `cftime.datetime`.
```

ledm commented:
@valeriupredoi, I've been doing all my development with `iris.__version__ = 1.13.0`, `matplotlib.__version__ = 2.2.2` and haven't seen this error before. 

I have replaced the Em dash ( –) in my recipes, but I've found that there are several other ones around the repository. I believe that it shouldn't matter in a yaml file if it's hidden in a comment.

valeriupredoi commented:
@ledm that's because you are working with obsolete versions of ```iris=1.13``` with ```nc_time_axis```, this problem has started to creep up only in the last two-three weeks, after you've started working on this PR; if you uninstall iris and nc_time_axis, remove the ```iris=1.13*``` and ```nc_time_axis*``` files from your ```anaconda/pkgs``` and download+reinstall both packages via conda you will see it happening. The preprocessor bit of that receipe works fine so I am happy on that side but you need to fix this issue before the PR gets merged, unfortunately

valeriupredoi commented:
to give you the repair solution, in ```/iris/plot.py``` line 425 should look like:

```
            r = [nc_time_axis.CalendarDateTime(
                 cftime.datetime(*date), coord.units.calendar)
                 for date in dates]
```
instead in iris 1.13, that same line looks like:
```
            r = [nc_time_axis.CalendarDateTime(
                 netcdftime.datetime(*date), coord.units.calendar)
                 for date in dates]
```
because nc_time_axis used to use netcdftime but not no more. Since we can not change the iris source cde, I suggest a workaround the plotting call -- either drop the quickplot, replace it with a matplotlib equivalent or produce an empty plot when using iris=<1.13. Note that this is fully solved by the dependency tree only in iris=2.1, so even in 2.0 you will have the same problemo

ledm commented:
Hi @valeriupredoi, it looks like the iris cftime issue is also linked with the time preproccessor, `time_slice()`, now in `preprocessor/_time_pp.py`, formally in `preprocessor/_time_area.py`, and also the  `preprocessor/_multimodel.py` function `_datetime_to_int_days`.

valeriupredoi commented:
@ledm what makes you think it affects it? it doesn't use ```netcdftime``` that now has changed to ```cftime``` in ```iris=2.1```

ledm commented:
Hi,
I think that there are some subtle differences between netcdftime and cftime. The time contraint in `time_slice` fails with the following error message:

```
2018-06-26 09:51:49,886 UTC [9911] ERROR   Program terminated abnormally, see stack trace below for more information
Traceback (most recent call last):
  File "/users/modellers/ledm/workspace/ESMValToolTest/ESMValTool/esmvaltool/_main.py", line 204, in run
    conf = main(args)
  File "/users/modellers/ledm/workspace/ESMValToolTest/ESMValTool/esmvaltool/_main.py", line 133, in main
    process_recipe(recipe_file=recipe, config_user=cfg)
  File "/users/modellers/ledm/workspace/ESMValToolTest/ESMValTool/esmvaltool/_main.py", line 183, in process_recipe
    recipe.run()
  File "/users/modellers/ledm/workspace/ESMValToolTest/ESMValTool/esmvaltool/_recipe.py", line 922, in run
    self.tasks, max_parallel_tasks=self._cfg['max_parallel_tasks'])
  File "/users/modellers/ledm/workspace/ESMValToolTest/ESMValTool/esmvaltool/_task.py", line 473, in run_tasks
    _run_tasks_sequential(tasks)
  File "/users/modellers/ledm/workspace/ESMValToolTest/ESMValTool/esmvaltool/_task.py", line 484, in _run_tasks_sequential
    task.run()
  File "/users/modellers/ledm/workspace/ESMValToolTest/ESMValTool/esmvaltool/_task.py", line 184, in run
    input_files.extend(task.run())
  File "/users/modellers/ledm/workspace/ESMValToolTest/ESMValTool/esmvaltool/_task.py", line 185, in run
    self.output_files = self._run(input_files)
  File "/users/modellers/ledm/workspace/ESMValToolTest/ESMValTool/esmvaltool/preprocessor/__init__.py", line 268, in _run
    input_files, self.settings, self.order, debug=self.debug)
  File "/users/modellers/ledm/workspace/ESMValToolTest/ESMValTool/esmvaltool/preprocessor/__init__.py", line 193, in preprocess_multi_model
    debug)
  File "/users/modellers/ledm/workspace/ESMValToolTest/ESMValTool/esmvaltool/preprocessor/__init__.py", line 230, in preprocess
    result.append(function(item, **args))
  File "/users/modellers/ledm/workspace/ESMValToolTest/ESMValTool/esmvaltool/preprocessor/_time_pp.py", line 44, in time_slice
    cube_slice = mycube.extract(my_constraint)
  File "/users/modellers/ledm/workspace/ESMValToolTest/miniconda/minicondaLee/envs/esmvaltool/lib/python3.6/site-packages/iris/cube.py", line 2223, in extract
    return constraint.extract(self)
  File "/users/modellers/ledm/workspace/ESMValToolTest/miniconda/minicondaLee/envs/esmvaltool/lib/python3.6/site-packages/iris/_constraints.py", line 151, in extract
    resultant_CIM = self._CIM_extract(cube)
  File "/users/modellers/ledm/workspace/ESMValToolTest/miniconda/minicondaLee/envs/esmvaltool/lib/python3.6/site-packages/iris/_constraints.py", line 175, in _CIM_extract
    resultant_CIM = resultant_CIM & coord_constraint.extract(cube)
  File "/users/modellers/ledm/workspace/ESMValToolTest/miniconda/minicondaLee/envs/esmvaltool/lib/python3.6/site-packages/iris/_constraints.py", line 290, in extract
    r = np.array([call_func(cell) for cell in coord.cells()])
  File "/users/modellers/ledm/workspace/ESMValToolTest/miniconda/minicondaLee/envs/esmvaltool/lib/python3.6/site-packages/iris/_constraints.py", line 290, in <listcomp>
    r = np.array([call_func(cell) for cell in coord.cells()])
  File "/users/modellers/ledm/workspace/ESMValToolTest/ESMValTool/esmvaltool/preprocessor/_time_pp.py", line 42, in <lambda>
    my_constraint = iris.Constraint(time=lambda t: (time1 < t.point < time2))
  File "cftime/_cftime.pyx", line 1606, in cftime._cftime.datetime.__richcmp__
TypeError: cannot compare cftime._cftime.DatetimeNoLeap(2001, 1, 16, 12, 0, 0, 0, 2, 16) and 18615.0

```
This is resolved by using the code reviously hidden behind a `TODO` statement:
```
        my_constraint = iris.Constraint(time=lambda t: (
        time1 < time_units.date2num(t.point) < time2)) 
```

Similarly, multimodel fails with:
```
Traceback (most recent call last):
  File "/users/modellers/ledm/workspace/ESMValToolTest/ESMValTool/esmvaltool/_main.py", line 204, in run
    conf = main(args)
  File "/users/modellers/ledm/workspace/ESMValToolTest/ESMValTool/esmvaltool/_main.py", line 133, in main
    process_recipe(recipe_file=recipe, config_user=cfg)
  File "/users/modellers/ledm/workspace/ESMValToolTest/ESMValTool/esmvaltool/_main.py", line 183, in process_recipe
    recipe.run()
  File "/users/modellers/ledm/workspace/ESMValToolTest/ESMValTool/esmvaltool/_recipe.py", line 922, in run
    self.tasks, max_parallel_tasks=self._cfg['max_parallel_tasks'])
  File "/users/modellers/ledm/workspace/ESMValToolTest/ESMValTool/esmvaltool/_task.py", line 473, in run_tasks
    _run_tasks_sequential(tasks)
  File "/users/modellers/ledm/workspace/ESMValToolTest/ESMValTool/esmvaltool/_task.py", line 484, in _run_tasks_sequential
    task.run()
  File "/users/modellers/ledm/workspace/ESMValToolTest/ESMValTool/esmvaltool/_task.py", line 184, in run
    input_files.extend(task.run())
  File "/users/modellers/ledm/workspace/ESMValToolTest/ESMValTool/esmvaltool/_task.py", line 185, in run
    self.output_files = self._run(input_files)
  File "/users/modellers/ledm/workspace/ESMValToolTest/ESMValTool/esmvaltool/preprocessor/__init__.py", line 268, in _run
    input_files, self.settings, self.order, debug=self.debug)
  File "/users/modellers/ledm/workspace/ESMValToolTest/ESMValTool/esmvaltool/preprocessor/__init__.py", line 201, in preprocess_multi_model
    debug)
  File "/users/modellers/ledm/workspace/ESMValToolTest/ESMValTool/esmvaltool/preprocessor/__init__.py", line 224, in preprocess
    result = [function(items, **args)]
  File "/users/modellers/ledm/workspace/ESMValToolTest/ESMValTool/esmvaltool/preprocessor/_multimodel.py", line 334, in multi_model_statistics
    interval = _get_overlap(selection)
  File "/users/modellers/ledm/workspace/ESMValToolTest/ESMValTool/esmvaltool/preprocessor/_multimodel.py", line 200, in _get_overlap
    span = _datetime_to_int_days(cube)
  File "/users/modellers/ledm/workspace/ESMValToolTest/ESMValTool/esmvaltool/preprocessor/_multimodel.py", line 168, in _datetime_to_int_days
    for cell in cube.coord('time').cells()]
  File "/users/modellers/ledm/workspace/ESMValToolTest/ESMValTool/esmvaltool/preprocessor/_multimodel.py", line 168, in <listcomp>
    for cell in cube.coord('time').cells()]
  File "/users/modellers/ledm/workspace/ESMValToolTest/miniconda/minicondaLee/envs/esmvaltool/lib/python3.6/site-packages/cf_units/__init__.py", line 1989, in num2date
    return _num2date_to_nearest_second(time_value, cdf_utime)
  File "/users/modellers/ledm/workspace/ESMValToolTest/miniconda/minicondaLee/envs/esmvaltool/lib/python3.6/site-packages/cf_units/__init__.py", line 601, in _num2date_to_nearest_second
    time_values % 1. == 0.5)
TypeError: unsupported operand type(s) for %: 'cftime._cftime.DatetimeNoLeap' and 'float'
```

valeriupredoi commented:
@ledm please don't change these lines -- the code labeled with ```TODO``` works only for ```iris => 2.0``` and this is exactly why that snippet stays there and is commented out, so we use it when we will be switching from 1.13. With ```iris 1.13``` it works as expected and there are no errors

ledm commented:
This feels a bit like a catch 22, @valeriupredoi. You told me to upgrade iris, and it broke several parts of the preprocessor and the `setup.py test` code, but now you're telling me I can't fix them. 

You also instructed me to add a version dependent code to my plotting tools, which I've done, but I can see whether it works, or whether I can share it as I'm not able to run the `setup.py test` or get past the preprocessor!

Also, apologies to everyone for adding so many code changes at once!

valeriupredoi commented:
hahahah @ledm sorry for confusing you, dude -- but what I said was rather clear: do not update the iris since we are still working with 1.13 in this release; change the just the plotting code that uses iris.quickplot (since that's the only bit affected by the transfer from ```netcdf``` to ```cftime```) in such a way that doean't break your diagnostic - how you do that is up to you entirely, what I did for the autoassess diag that uses quickplot as well: I put a check on the iris version - if = 2.1 execute the plotting, if not, return an empty plot

valeriupredoi commented:
hey @ledm man, cheers for making the needed changes, now the namelist is running fine! A bit on profiling, I have had the namelist crash a couple times with `MemoryError` and have noticed that certain diags may take up to 18G of memory. The functions that affected this are:
```
_volume_pp:
  depth_integration(): iris.analysis.SUM
  volume_average(): iris.analysis.MEAN
```
If I were you I'd be careful with such things because a user may be running the diag on a node that hasn't much memory (note that I ran with `max_parallel_tasks:1` in config). Where possible, don't rely on the iris.analysis since that one is not playing well in terms of resources with large numerical arrays. Anyhoo, the thing is still running now (the third attempt) and albeit memory still went to 14G a few times, Jasmin sci1 seems to cope with it :grin:

ledm commented:
Thanks a lot for all the help and advice, @valeriupredoi, @jvegasbsc, @bouweandela, @mattiarighi, and @koldunovn. As far as I can tell, I've addressed all the comments that you've given so far, @valeriupredoi has tested the diagnostic scripts. From here, it looks like the merge pull request is ready to go. 

There are still quality review failures, but there's little that I can do about the remaining Codacy issues. The remaining issues are things like:
- codacy wants the `allvars` class n the `_fixes` files  to be capitalised, 
- codacy not being able to import various modules (like iris), 
- Actual real problems with functions that I did not author, like `trajectory_cube` in `preprocessor/_area_pp.py` (formely in `preprocessor/_time_area.py`). 

The problems with Memory that @valeriupredoi mentioned could be addressed by iterating these functions over the `time` or `plev` dimension, but neither of us think that this is an essential task needed before the merge can be completed.

mattiarighi commented on esmvaltool/preprocessor/__init__.py:
```diff
@@ -47,8 +53,14 @@
     'average_region',
+    'average_volume',
+    'time_average',
     # 'average_zone': average_zone,
```
For the reasons outlined in the discussion, I think this should be moved after `multi_model_statistics`, to keep the processing chain as general as possible.

mattiarighi commented on esmvaltool/preprocessor/__init__.py:
```diff
@@ -47,8 +53,14 @@
+    'depth_integration',
     'average_region',
+    'average_volume',
+    'time_average',
```
For the reasons outlined in the discussion, I think this should be moved after `multi_model_statistics`, to keep the processing chain as general as possible.

valeriupredoi replied:
the `_multimodel` module should work fine with a single time point but will otherwise fail if there is no time axis. I agree with @mattiarighi on this one for generality purposes; having said this, if the multimodel statistics is intended to be computed for a time-averaged model ensemble (whether it be from daily data to monthly averages, from monthly data to yearly averages etc) then it is probably desirable to apply the time operation first and keep the time axis with a certain number of points and then apply the multimodel computation, that would save a lot of resources

mattiarighi replied:
I agree that resources could be an issue, but it's conceptually not correct to apply time/area statistics before the multi-model. This is because, in general, the multi-model of some stats is not equal to the stats of the multi-model. 

Since the multi-model-mean/median is treated like another dataset, it's important that the preprocessing is applied consistently to all of them.

That's why I'm insisting on keeping the order of operations as originally planned. I understand that some diagnostic developers (like @ledm in this PR) would like to customize this order, but as I said this is not yet support and until then we need to stay with the most general approach.

valeriupredoi commented:
Yup, the old f(g(x)) problem - very good point that I missed @mattiarighi

Dr Valeriu Predoi.
Computational scientist
NCAS-CMS
University of Reading
Department of Meteorology
Reading RG6 6BB
United Kingdom

On Fri, 29 Jun 2018, 07:45 Mattia Righi, <notifications@github.com> wrote:

> *@mattiarighi* commented on this pull request.
> ------------------------------
>
> In esmvaltool/preprocessor/__init__.py
> <https://github.com/ESMValGroup/ESMValTool/pull/438#discussion_r199066818>
> :
>
> >      'average_region',
> +    'average_volume',
> +    'time_average',
>
> I agree that resources could be an issue, but it's conceptually not
> correct to apply time/area statistics before the multi-model. This is
> because, in general, the multi-model of some stats is not equal to the
> stats of the multi-model.
>
> Since the multi-model-mean/median is treated like another dataset, it's
> important that the preprocessing is applied consistently to all of them.
>
> That's why I'm insisting on keeping the order of operations as originally
> planned. I understand that some diagnostic developers (like @ledm
> <https://github.com/ledm> in this PR) would like to customize this order,
> but as I said this is not yet support and until then we need to stay with
> the most general approach.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/ESMValGroup/ESMValTool/pull/438#discussion_r199066818>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AbpCo9wnQTRoXPgPVfQCnASu1qzaWUlLks5uBcz6gaJpZM4Utv3M>
> .
>

ledm commented:
I'm convinved too. I've moved the time average after the multimodel
preprocessor and pushed the change.

Please let me know if you spot anything else to do,

Lee


Please visit our new website at www.pml.ac.uk and follow us on Twitter @PlymouthMarine

We have updated our Privacy Notice which is available at http://www.pml.ac.uk/System-pages/Privacy_Notice

Plymouth Marine Laboratory (PML) is a company limited by guarantee registered in England & Wales, company number 4178503. Registered Charity No. 1091222. Registered Office: Prospect Place, The Hoe, Plymouth  PL1 3DH, UK. 

This message is private and confidential. If you have received this message in error, please notify the sender and remove it from your system. You are reminded that e-mail communications are not secure and may contain viruses; PML accepts no liability for any loss or damage which may be caused by viruses.

mattiarighi commented:
Thanks @ledm.

I would also move the other grid-point operations (`depth_integration`, `average_region`, `average_volume`) and `seasonal_mean` after multi-model, for the same reasons.

The extract operations can stay before, they should be not critical in this sense (extracting before or after multi-model does not affect the results, but it saves memory)

valeriupredoi commented:
@mattiarighi I think so, on my way to work and will double check when I get
there

Dr Valeriu Predoi.
Computational scientist
NCAS-CMS
University of Reading
Department of Meteorology
Reading RG6 6BB
United Kingdom

On Fri, 29 Jun 2018, 09:26 Mattia Righi, <notifications@github.com> wrote:

> Sorry, I didn't notice that season_mean was already there.
> @valeriupredoi <https://github.com/valeriupredoi> where does it come
> from? Was it part of your very first implementation of time-area?
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/ESMValGroup/ESMValTool/pull/438#issuecomment-401285818>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AbpCowFk3qpbdhNyBnkBHbugohgMQ5vCks5uBeSygaJpZM4Utv3M>
> .
>

valeriupredoi commented:
@mattiarighi sorry, being a big mouth I just ran into Grenville and had to give him a very detailed updated on ESMValTool :grin: yes, the `seasonal_mean()` from `_time_area.py` is a function I put in about a year ago, but it still seems to do the trick OK, @ledm are you using it or you've implemented your own seasons?

ledm commented:
I haven't looked at the seasonal mean function at all. I hate to speak for everyone, but ocean scientists are typically are happy with monthly or annual means.

I should also point out that the `time_average` is an extremely simple function which collapses the cube along the time dimension. Perhaps it would be worth weighting the average according to the length of each time step, in the case of uneven time dimensions, but it doesn't do that yet.

valeriupredoi commented:
yea no probs @ledm - I am autoassess-assing now and the land-atmos diags need the seasons; anyway, the current multimodel stats function is not applying weights (yet) so I reckon you don't need weights yourself just yet. Let's not postpone more this PR just now, it's already a lot of stuff that needs to get in anyway :smiling_imp:

mattiarighi commented on esmvaltool/preprocessor/_multimodel.py:
```diff
@@ -73,7 +73,10 @@ def _compute_statistic(datas, name):
         # datas is per time point
         # so we can safely NOT compute stats for single points
-        u_datas = [data for data in datas if not np.all(data.mask)]
+        if datas.ndim == 1:
```
Do we still need all this and the cases below, now that we moved time-area after multi-model?

mattiarighi replied:
Alright, let's leave it there :+1:

mattiarighi commented on esmvaltool/preprocessor/_multimodel.py:
```diff
@@ -113,12 +116,20 @@ def _put_in_cube(template_cube, cube_data, stat_name,
+        # If you're going to hardwire air_pressure into this,
+        # might as well have depth here too.
+        plev = template_cube.coord('depth')
+        cspec = [(times, 0), (plev, 1), ]
```
See above comment.

ledm replied:
Hi, to be honest, this whole section could use some work (outside of this PR). For one, it appears that both the 'air_pressure' and 'depth' dimension names are hardwired into the preprocessor. Also, it appears that the four conditions are only linked with the number of dimensions, instead of which dimensions are present in the cube.

I'm guessing @valeriupredoi may want to address this in the future?

bouweandela commented on esmvaltool/diag_scripts/ocean/diagnostic_timeseries.py:
```diff
@@ -0,0 +1,280 @@
+        # Plot each file in the group
+        for index, filename in enumerate(sorted(metadata)):
+            color = cmap(
+                (float(index) / (len(metadata.keys()) - 1.)))
```
slightly nicer code: `color = cmap(index / (len(metadata) - 1.))`

bouweandela commented on esmvaltool/diag_scripts/ocean/diagnostic_timeseries.py:
```diff
@@ -0,0 +1,280 @@
+            color = cmap(
+                (float(index) / (len(metadata.keys()) - 1.)))
+
+            if metadata[filename]['dataset'].find('MultiModel') > -1:
```
nicer code: `if 'MultiModel' in metadata[filename]['dataset']:`

bouweandela commented on esmvaltool/diag_scripts/ocean/diagnostic_tools.py:
```diff
@@ -0,0 +1,207 @@
+    Returns folder string and final character is always '/'.
+    """
+    if isinstance(name, list):
+        name = '/'.join(name)
```
Use `os.sep` instead of hardcoding `/`

bouweandela commented on esmvaltool/preprocessor/_area_pp.py:
```diff
@@ -0,0 +1,110 @@
+
+
+# slice cube over a restricted area (box)
+def area_slice(mycube, long1, long2, lat1, lat2):
```
Use names that are easier to read, e.g.:
```
def area_slice(cube, longitude, latitude):
```
with `latitude` and `longitude` lists with two items in them, start and end, or
``` 
def area_slice(cube, start_longitude, end_longitude, start_latitude, end_latitude):
```
if you want to be more explicit

bouweandela commented on esmvaltool/preprocessor/_area_pp.py:
```diff
@@ -0,0 +1,110 @@
+
+
+# get zonal means
+def zonal_means(mycube, coord1, mean_type):
```
Use names that are easy to read, e.g.:
`def zonal_means(cube, coordinate, mean_type):`

bouweandela commented on esmvaltool/preprocessor/_regrid.py:
```diff
@@ -23,8 +21,11 @@
-_CELL_SPEC = re.compile(r'''\A
+_CELL_SPEC = re.compile(
+    r'''\A
                             \s*(?P<dx>\d+(\.\d+)?)\s*
```
align the string so it is easy to read again: yapf does not change the content of strings

bouweandela commented on esmvaltool/preprocessor/_regrid.py:
```diff
@@ -53,7 +54,7 @@
 
 # Supported vertical interpolation schemes.
-vertical_schemes = ['linear', 'nearest']
+vertical_schemes = ['linear', 'nearest', 'nearest_extrap', 'linear_extrap']
```
Use easy to read names, e.g. 'nearest_neighbour_extrapolation` instead of 'nearest_extrap'

ledm replied:
Went with the explicit naming scheme "linear_horizontal_extrapolate_vertical" to describe the new schemes.

bouweandela commented on esmvaltool/preprocessor/_volume_pp.py:
```diff
@@ -0,0 +1,170 @@
+    return region_subset
+
+
+def volume_average(mycube, coordz, coord1, coord2,):
```
Remove the , after coord2

bouweandela commented on esmvaltool/recipes/recipe_OceanBGC.yml:
```diff
@@ -0,0 +1,132 @@
+###############################################################################
```
Please use the tool `yamllint` to check that your yaml file looks as good as possible, you can ignore any comments about lines too long or too much whitespace if you think your layout is nicer than what yamllint suggests.

bouweandela commented on esmvaltool/recipes/recipe_OceanBGC.yml:
```diff
@@ -0,0 +1,132 @@
+  # Depth-profile preprocessors
+  # --------------------------------------------------
+  prep_depth_integration_1: # For a 2D map global profile
+    fix_data:     
```
Do you need the empty `fix_data` and `fix_metadata` dicts? Fixes should be applied fully automatically by the preprocessor and there should be no need to specify them here.

bouweandela replied:
@ledm Can you have a look at this?

bouweandela commented on esmvaltool/recipes/recipe_OceanPhysics.yml:
```diff
@@ -0,0 +1,502 @@
+        preprocessor: prep_depth_integration_1      
+        field: TO3M
+    scripts:
+      Global_Ocean_DepthIntegration_map: &Global_Ocean_DepthIntegration_map
```
Remove the anchor `&Global_Ocean_DepthIntegration_map`, it looks like you're not using that anywhere. See [here](https://blog.daemonl.com/2016/02/yaml.html) for a short explanation of YAML anchors.

bouweandela commented:
See e.g. the file `tests/integration/preprocessor/_regrid/test_regrid.py` for examples of how to implement tests.

ledm commented:
Hi @bouweandela, thanks for the extensive review comments, they are very helpful. I've address most of these points now and I'll be submitting these changes in the near future. I have yet to start adding tests, but I'll see what I can do on that front. 

One issue I'm seeing is that I'm finding that the reordering the preprocessor operations in `__init__.py`  breaks several aspects of the ocean analysis [(commit info)](https://github.com/ESMValGroup/ESMValTool/pull/438/commits/93ae6d859785f114548010d5d8468e7d549ec610). 
The region_average and volume_average reduce a 2 or 3 dimensional array to 1D array (with dimension: `time`), which works fine if run before the multi-model mean. If we try to run these preproccessors afterwards, then the multi-model mean will try to take the mean of several models which all have different model grids shapes, and it fails. 

This means that we now need to regrid everything in order to perform any kind of multi-model comparison of area or volume averages. As we all know, regridding can introduce uncertainties and increased running time, and it would be better to keep the average_volume and average_area preprocessors based on their naive data grids (before the multimodel preprocessor). Sorry to open this up again @mattiarighi and @valeriupredoi.

mattiarighi commented:
Fine with me to introduce customizable order of operation now, if it is not too much work.

But I think this would require to also introduce some checks to make sure the different modules of the preprocessor are interfacing correctly. Or, as @bouweandela suggests, we leave this to the user at his/her own risk.

bouweandela commented:
I've created an issue for this feature request here: #495.

ledm commented:
With regards to the ordering, I've also found that putting the `area_average`, `volume_average` and `time_average` preprocessor after the multi-model preprocessor means that the multi-model dataset hasn't had the same manipulations as the other models. Ie, in my case, multi-model data still contains all time points, where as the rest of the models are a time average. Similarly, the `volume_average` should have time-only dimensional, where as the multi-model contains latitude, longitude, time and depth dimensions.

This means that without re-ordering the preprocessors, very few of the ocean temperature examples will work (with multimodel data).

bouweandela commented:
Created issue #497 from the above comment.

ledm commented:
Sorry, I'm working on some early tests and accidentally pushed a draft commit, but it's far from complete.

Also, I'm adding unit tests instead of integration tests as these are more suited to test individual preprocessors.

ledm commented:
Hi @bouweandela, 

I've added several unit tests, and luckily these highlighted some conceptual issues with the preprocessor code which have now been fixed. Thanks for asking me to do them - it has resulted in significant improvements! 

Bearing in mind issue #495 and #497 (which currently mean that the standard preprocessing order does not work with the ocean temperature yml recipe), I think that this code is now ready for the PR. 

However, there may well be some problems as this PR might conflict with @jvegasbsc's PR #500, as we have split the _time_area.py into two separate files. 

(Also, I won't be able to work on ESMValTool until the end of next week, so fingers crossed that this is sufficient to get it over the line. )

Lee

bouweandela commented:
Hi @ledm,

Thanks for adding the tests, looks really good!

Could you also add [numpy style docstrings](http://sphinxcontrib-napoleon.readthedocs.io/en/latest/example_numpy.html) to the preprocessor functions that you added?

I noticed that you cleaned up one of the two namelists, but not the other one (recipe_OceanBGC.yml). Does that one work yet?

Could you also add a `documentation` section to your namelists? See .e.g recipe_perfmetrics_CMIP5.yml for an example. You can do this by moving the stuff that is in the comments at the top of the namelist into the proper yaml tags (add any tags used to esmvaltool/config-references.yml as well). This does not do anything yet, but when I get round to implementing provenance next month it will be used for that, see #240 for more information.

ledm commented:
Hi, I've added longer documentation for the preprocessors (though I must say that very few of the other preprocessors had this level of detail.)

I've also added the documentation section to the two recipes. As provenance is no yet implemented, I've just added a rough outline, but it should be sufficient for now. I'm happy to go into more depth when that part of the code is implemented. Either way, I intend to continue to add new part to these recipes and other recipes.

bouweandela commented on esmvaltool/recipes/recipe_OceanBGC.yml:
```diff
@@ -0,0 +1,117 @@
+  # map preprocessors
+  # --------------------------------------------------
+  prep_surface_map_1: # For a 2D global surface map
+    fix_data:     
```
Same here?

ledm replied:
Can you be more specific as to what the problem is, please?

bouweandela replied:
Do you need the empty `fix_data` and `fix_metadata` dicts? Fixes should be applied fully automatically by the preprocessor and there should be no need to specify them here.

bouweandela commented on esmvaltool/recipes/recipe_OceanPhysics.yml:
```diff
@@ -0,0 +1,451 @@
+      statistics: [mean ]
+
+  prep_timeseries_2: # For specific levels of 3D fields
+    fix_data:
```
And here ..

bouweandela replied:
Do you need the empty `fix_data` and `fix_metadata` dicts? Fixes should be applied fully automatically by the preprocessor and there should be no need to specify them here.

bouweandela commented on esmvaltool/recipes/recipe_OceanPhysics.yml:
```diff
@@ -0,0 +1,451 @@
+      statistics: [mean ]
+
+  prep_timeseries_7: # For regional volume averaged (at the surface)
+    fix_data:
```
And here ..

bouweandela replied:
Do you need the empty `fix_data` and `fix_metadata` dicts? Fixes should be applied fully automatically by the preprocessor and there should be no need to specify them here.

bouweandela commented:
Thank you for adding the documentation. You are right that the documentation for many of the other preprocessor functions also needs to be updated, we have an open issue on that #360. In order not the make the issue worse, I asked you to already provide it now. Could you check that your formatting is according to the numpy docstring style? At the moment it looks e.g. like this:
```
Arguments
---------
    cube: input cube.
```
while it should be:
```
Arguments
---------
cube : iris.cube.Cube
    Input cube.
```
As shown in the examples here: http://sphinxcontrib-napoleon.readthedocs.io/en/latest/example_numpy.html

bouweandela commented:
Thanks for fixing the documentation.

Regarding the `main` functions in your scripts: instead of reading in the variable metadata yourself, you can simply use the metadata provided in `cfg['input_data']`. If you would like to group and/or sort it, you can now (since PR #508) use the function `group_metadata`, see for a simple example esmvaltool/diag_scripts/examples/diagnostic.py.

valeriupredoi commented:
hey fellas @bouweandela @mattiarighi @ledm cheers for all your inputs and great work on this one -- I strongly suggest to approve this PR before suggesting any more changes -- it is already blowing up to humongous size and I feel that if suggestions keep coming this will be a neverending story, since the code base also keeps changing. Lee's effort is needed in since this will be a very useful set of diagnostics that a lot of scientists are already asking (at lest) me about. Lee can always branch out after this gets in and make the other changes needed to align the diags with the new implementations. I had a look through the Codacy issues and they are fine, nothing more than what I've had when merging the first Autoassess PR. In the future, as it was the case with the multimodel PR, I strongly suggest we limit the PR size and plug things in incrementally.

valeriupredoi commented:
yo guys, any opinions on merging this now and letting Lee work further on it on the branch? :grin:

ledm commented:
I've made a little push with minor changes, and I've commented out all the diagnostics that don't work without the custom preprocessor order. I

ledm commented:
Hi, I've just merged the latest version2_development branch into this pull. The custom order ocean temperature is now fully functional and tested. I think that the pull request is ready to complete.

Also note that the failures in Codacy are beyond the scope of this PR.

ledm commented:
Can we please complete this pull request @valeriupredoi @bouweandela. Everything worked yesterday and was ready to go. Ten this morning, there was two conflicting files, so I just pulled in the latest changes and now I have 22 test errors. 

The underlying development branch continues to move on and I can't afford the time to merge, debug, test, run and resubmit this.

bouweandela commented:
Hi Lee,

I'm fine with merging. The failing tests are probably because you made a mistake with the merge. Javier recently added some time operations to the preprocessor and also tests for those, that's what's causing the conflict (and probably your test failures). I performed the merge for you and uploaded it in the `REFACTORING_OceanTemperature2` branch, because I didn't want to mess with your branch. Maybe you can check if that looks good/works and then we can merge that one? I think someone independent from you should also test that your recipes actually work. @mattiarighi, did you already have a look?

ledm commented:
Thanks, the REFACTORING_OceanTemperature2 seems to work fine for me. I know that @valeriupredoi has been able to run this before.

ledm commented:
Closing this PR and recreated it as #526.
